api/handlers.go

```
@@ -165,3 +171,404 @@ func (api *RecipeAPI) AddRecipeHandler(w http.ResponseWriter, req *http.Request)
+	if err != nil {
+		log.Event(ctx, "error updating recipe to mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Marshal to output the newly added recipe
+	output, err := json.Marshal(currentRecipe)
+	if err != nil {
+		log.Event(ctx, "error returned from json marshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("content-type", "application/json")
+	w.Write(output)
+}
+
+//UpdateRecipeHandler - update specific recipe by ID
+func (api *RecipeAPI) UpdateRecipeHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	//Get Update Recipe ID
+	vars := mux.Vars(req)
+	recipeID := vars["id"]
+	logD := log.Data{"recipe_id": recipeID}
+
+	// Read body
+	b, err := ioutil.ReadAll(req.Body)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Response struct
+	var updates recipe.Response
+	err = json.Unmarshal(b, &updates)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	//Validation to check if all the recipe fields are entered
+	err = updates.ValidateUpdateRecipe()
+	if err != nil {
+		log.Event(ctx, "bad request error for invalid updates given in request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Update Recipe to Mongo
+	update := bson.M{"$set": updates}
+	err = api.dataStore.Backend.UpdateRecipe(recipeID, update)
+	if err != nil {
+		log.Event(ctx, "error updating recipe to mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Marshal to output the newly updated recipe
+	updatedRecipe, _ := api.dataStore.Backend.GetRecipe(recipeID)
```

never discard the error, always handle it please

---

Done in commit "move bson statements to mongo package - mongo package encapsulation"